src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala

```diff
@@ -1,126 +1,198 @@
+              } yield {
+                entityWithId.addUrl().removeIds()
+              }
+            }
+          }
+        }
+      }
+    }
+  }
 
+  def delete(agoraEntity: AgoraEntity, entityTypes: Seq[AgoraEntityType.EntityType], username: String): Future[Int] = {
+    //list of associated configurations. Goes to Mongo so we do it outside the sql txn
+    val configurations = if (entityTypes equals AgoraEntityType.MethodTypes) {
       val dao = AgoraDao.createAgoraDao(entityTypes)
       val entityWithId = dao.findSingle(agoraEntity.namespace.get, agoraEntity.name.get, agoraEntity.snapshotId.get)
-      val configurations = dao.findConfigurations(entityWithId.id.get)
-
-      configurations.foreach {config => AgoraEntityPermissionsClient.deleteAllPermissions(config)}
+      dao.findConfigurations(entityWithId.id.get)
+    } else {
+      Seq()
     }
 
-    AgoraEntityPermissionsClient.deleteAllPermissions(agoraEntity)
+    permissionsDataSource.inTransaction { db =>
+      db.aePerms.addUserIfNotInDatabase(username) flatMap { _ =>
+        checkNamespacePermission(db, agoraEntity, username, AgoraPermissions(Redact), checkAdmin = true) {
+          //admins can redact anything
+          DBIO.sequence(configurations map { config => db.aePerms.deleteAllPermissions(config) }) andThen
+            db.aePerms.deleteAllPermissions(agoraEntity)
+        }
+      }
+    }
   }
 
   def find(agoraSearch: AgoraEntity,
            agoraProjection: Option[AgoraEntityProjection],
            entityTypes: Seq[AgoraEntityType.EntityType],
-           username: String): Seq[AgoraEntity] = {
+           username: String): Future[Seq[AgoraEntity]] = {
 
     val entities = AgoraDao.createAgoraDao(entityTypes)
       .find(agoraSearch, agoraProjection)
       .map(entity => entity.addUrl().removeIds())
 
-    AgoraEntityPermissionsClient.filterEntityByRead(entities, username)
+    permissionsDataSource.inTransaction { db =>
+      for {
+        _ <- db.aePerms.addUserIfNotInDatabase(username)
+        entity <- db.aePerms.filterEntityByRead(entities, username)
+      } yield entity
+    }
   }
 
   def findSingle(namespace: String,
                  name: String,
                  snapshotId: Int,
                  entityTypes: Seq[AgoraEntityType.EntityType],
-                 username: String): AgoraEntity = {
+                 username: String): Future[AgoraEntity] = {
     val foundEntity = AgoraDao.createAgoraDao(entityTypes).findSingle(namespace, name, snapshotId)
-    val entities = AgoraEntityPermissionsClient.filterEntityByRead(Seq(foundEntity), username)
-    entities match {
-      case Seq(ae: AgoraEntity) => ae.addUrl().removeIds().addManagers(AgoraEntityPermissionsClient.listOwners(foundEntity))
-      case _ => throw new AgoraEntityNotFoundException(foundEntity)
+
+    permissionsDataSource.inTransaction { db =>
+      db.aePerms.addUserIfNotInDatabase(username) flatMap { _ =>
```

Comment:
I removed a lot of calls to `addUserIfNotInDatabase` because they were happening in a separate txn. they live where they should now!